x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDatePeriod.java

```diff
@@ -0,0 +1,162 @@
+        "day",
+        "days",
+        "d",
+        "week",
+        "weeks",
+        "w",
+        "month",
+        "months",
+        "mo",
+        "quarter",
+        "quarters",
+        "q",
+        "year",
+        "years",
+        "yr",
+        "y"
+    );
+
+    private static final String INVALID_INTERVAL_ERROR = "Invalid interval value in [{}], expected integer followed by one of "
+        + intervals
+        + " but got [{}]";
+
+    @FunctionInfo(
+        returnType = "date_period",
+        description = "Converts an input value into a date_period value.",
+        examples = @Example(description = "Converts an input value into a date_period value.", file = "convert", tag = "castToDatePeriod")
+    )
+    public ToDatePeriod(
+        Source source,
+        @Param(name = "field", type = { "date_period", "keyword", "text" }, description = "Input value.") Expression v
+    ) {
+        super(source, v);
+    }
+
+    private ToDatePeriod(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected Map<DataType, BuildFactory> factories() {
+        return EVALUATORS;
+    }
+
+    @Override
+    protected final TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        TypeResolution typeResolution = isType(
+            field(),
+            factories()::containsKey,
+            sourceText(),
+            FIRST,
+            supportedTypesNames(supportedTypes())
+        );
+        if (field().foldable()) {
+            if (isString(field().dataType()) && isValidInterval(field().fold().toString()) == false) {
+                typeResolution = new TypeResolution(LoggerMessageFormat.format(null, INVALID_INTERVAL_ERROR, sourceText(), field().fold()));
+            }
+
+        }
+        return typeResolution;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DATE_PERIOD;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new ToDatePeriod(source(), newChildren.get(0));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, ToDatePeriod::new, field());
+    }
+
+    @Override
+    public final Object fold() {
+        if (field() instanceof Literal l) {
+            Object v = l.value();
+            if (v instanceof BytesRef b) {
+                if (isValidInterval(b.utf8ToString()) == false) {
+                    throw new IllegalArgumentException(
+                        LoggerMessageFormat.format(null, INVALID_INTERVAL_ERROR, sourceText(), b.utf8ToString())
+                    );
+                }
+                return EsqlDataTypeConverter.parseTemporalAmount(b.utf8ToString(), DATE_PERIOD);
+            }
+        }
+        return null;
```

Comment:
I don't think this is correct here... not sure, but I get a NPE if I run `"from employees | EVAL x = birth_date + first_name::date_period`

Reply:
> from employees | EVAL x = birth_date + first_name::date_period

These two conversion functions don't support a field as input yet(at least in this PR, it doesn't work), it only takes literals, as we don't have a Block type that matches date_period or time_duration, need to add more checks here.